### PR TITLE
Bug 1273034 - Transition to Taskcluster Auth from Persona

### PR DESCRIPTION
**Pull request by camd**
Switch to using Taskcluster auth instead of Persona directly.
Still possible to use Persona while login.taskcluster.net supports it.

Bug 1315818 to use localhost instead of local.treeherder.mozilla.org is in another PR #1974

Not included in this PR:

* Bug 1315826 - Logging in directly in the Django admin
    You are no longer able to login directly to the Django admin in this commit.  You must first login     through the Treeherder main ui, then switch to the Django admin and use it normally.  You'll see a message to that effect on the admin screen if you're not logged in.

* Bug 1273092 - to use LDAP scopes to determine Sheriff access

---

This change is [Reviewable](https://reviewable.io/reviews/mozilla/treeherder/1922)

### CODE REVIEW

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

File: treeherder/config/settings.py [hunk 1]
```diff
@@ -102,14 +102,9 @@ def server_supports_tls(url):
 
 AUTHENTICATION_BACKENDS = (
     'django.contrib.auth.backends.ModelBackend',
-    'django_browserid.auth.BrowserIDBackend',
+    'django.contrib.auth.backends.RemoteUserBackend',
```
Comment (edmorley) [inline treeherder/config/settings.py]: I'm pretty sure we need to be using login() from here:
https://docs.djangoproject.com/en/1.8/topics/auth/default/#how-to-log-a-user-in

And not the `RemoteUserBackend` function, which as far as I can tell is meant for when the `REMOTE_USER` HTTP header is set by a service in front of the app (eg an LDAP apache plugin in a corporate environment).

File: treeherder/webapp/api/auth.py [hunk 2]
```diff
@@ -16,3 +27,61 @@ def hawk_lookup(id):
...
+            authenticated = authenticate(remote_user=user.username)
+            login(request, authenticated)
+
+            return Response({
+                "message": "login success",
+                "result": str(result),
+                "user": {"email": user.email,
+                         "isSheriff": user.is_staff}
+            })
+        else:
```
Comment (edmorley) [inline treeherder/webapp/api/auth.py]: Unnecessary else.

File: treeherder/webapp/api/auth.py [hunk 3]
```diff
@@ -16,3 +27,61 @@ def hawk_lookup(id):
...
+            return Response({"message": "login failed",
```
Comment (edmorley) [inline treeherder/webapp/api/auth.py]: I'd ask @djmitche if there is ever anything sensitive in the result that we shouldn't be returning to the client.

File: ui/admin.html [hunk 4]
```diff
@@ -70,7 +69,7 @@
   <script src="vendor/angular/angular-cookies.js"></script>
   <script src="vendor/angular/angular-ui-router.js"></script>
   <script src="vendor/angular/angular-sanitize.js"></script>
-  <script src="vendor/angular-local-storage.min.js"></script>
+  <script src="vendor/ngStorage.min.js"></script>
```
Comment (edmorley) [inline ui/admin.html]: Why this change? If it's just a better/newer library for localstorage, can we make it as part of a separate PR that can land sooner?

File: ui/partials/admin/exclusions_list.html [hunk 5]
```diff
@@ -3,8 +3,8 @@
        ng-controller="ExclusionsListCtrl"
        ng-init="populateExclusionsData()">
     <button ng-click="go('exclusions_detail', {id: 'new'})" type="button"
-            ng-disabled="!user.is_staff"
```
Comment (edmorley) [inline ui/partials/admin/exclusions_list.html]: I think we should either hold off the s/is_staff/isSheriff/ until after this bug, or do so in an earlier PR to make the diff easier to read.

**Comment by edmorley** [comment]
This may be useful:
https://docs.djangoproject.com/en/1.8/topics/auth/default/#module-django.contrib.auth.views

File: treeherder/config/settings.py (hunk 1 above)
Comment (camd) [inline treeherder/config/settings.py]: Yeah, agreed about the custom backend.  I had thought about that yesterday as well.  Thanks for taking a look at this while it's still pretty messy.  My comment in IRC, I somehow missed the part where you said "I'm pretty sure we need to be using a custom backend"  So my comment in IRC probably seemed like I was looney...  :)

File: treeherder/webapp/api/auth.py (hunk 3 above)
Comment (camd) [inline treeherder/webapp/api/auth.py]: agreed, yeah.  this was also for debugging.  I won't return the result.

File: ui/admin.html (hunk 4 above)
Comment (camd) [inline ui/admin.html]: this is an easier package to use and we weren't using localstorage anywhere else, afaict.  But, yeah, I'll move it into another PR.

File: ui/partials/admin/exclusions_list.html (hunk 5 above)
Comment (camd) [inline ui/partials/admin/exclusions_list.html]: agreed, yeah.  I'll do that later.

File: treeherder/auth/backends.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+                "authorization": authorization,
+                "host": host,
+                "port": port,
+                "resource": "/",
+                "method": "get",
+            })
+
+            if result["status"] == "auth-success":
+                username = result["clientId"][-30:]
+                email = result["clientId"].split("/")[1]
```
Comment (djmitche) [inline treeherder/auth/backends.py]: This is assuming a clientId of the form `<something>/<email>` which happens to match the temporary credentials that taskcluster-login hands out, but not all clientIds have that form; see the full list of permanent credentials (which can be shared just as easiliy) at https://tools.taskcluster.net/auth/clients/

ClientIds can be used as visual feedback ("Hello, dmitchell@mozilla.com!"), but should not be used for access control ("Our records show that dmitchell@mozilla.com has permission to ..").  It's a little hard to tell which is the case here?

File: treeherder/webapp/api/auth.py [hunk 7]
```diff
@@ -16,3 +26,30 @@ def hawk_lookup(id):
...
+        host = request.query_params.get("host", None)
+        port = request.query_params.get("port", None)
+
+        user = authenticate(authorization=authorization, host=host, port=int(port))
+        login(request, user)
+
+        return Response({
+            "message": "login success",
+            "user": {"email": user.email,
+                     "is_staff": user.is_staff}
```
Comment (djmitche) [inline treeherder/webapp/api/auth.py]: Ah, so there's a bit of authorization going on here based on the clientId :(

Could we base that on a scope instead?  You could just check whether the user has the `assume:project:treeherder:staff` scope in the response from `authenticateHawk`.  It'd be fine to reflect that into the User object on every login, if you're comfortable with a user's staff-ness persisting until their TH login expires, even if they lose the scope (for example through an LDAP group change or becoming ex-staff)

File: ui/js/components/auth.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        );
+
+        // send a request from client side to TH server signed with TC
+        // creds from login.taskcluster.net
+        $http.get(`${urlBase}api/auth/login/?host=${host}&port=${port}`,
+                  {headers: {"oth": header.field}})
+            .then(function(resp) {
+                $localStorage.user = resp.data.user;
+                $window.close();
+            });
```
Comment (djmitche) [inline ui/js/components/auth.js]: This looks great!

File: treeherder/auth/backends.py [hunk 9]
```diff
@@ -2,13 +2,52 @@
...
+            if scope.startswith(scope_prefix):
+                return scope[len(scope_prefix):]
+        return None
+
+    def _get_email(self, result):
+        """
+        Get the user's email from the mozilla-user scope
+
+        If that scope is not present, then don't log the user in.  Cameron
+        and Dustin decided this should be a hard requirement.
```
Comment (djmitche) [inline treeherder/auth/backends.py]: You might want to link to https://docs.taskcluster.net/manual/3rdparty here

File: treeherder/auth/backends.py [hunk 10]
```diff
@@ -24,27 +63,39 @@ def authenticate(self, authorization=None, host=None, port=None):
...
-                                username=username,
+                                username=username[:30],
                                 password=sha.hexdigest()[25:]
                                 )
-                    user.save()
 
+                # User's sheriff status is no longer fetched from the DB.  Now
+                # it is determined solely by having the correct scope.
+                user.is_staff = scope_match(
+                    result["scopes"], [["assume:project:treeherder:sheriff"]])
```
Comment (djmitche) [inline treeherder/auth/backends.py]: :+1:

File: treeherder/auth/backends.py [hunk 11]
```diff
@@ -2,13 +2,52 @@
...
+
+    def _get_email(self, result):
+        """
+        Get the user's email from the mozilla-user scope
+
+        If that scope is not present, then don't log the user in.  Cameron
+        and Dustin decided this should be a hard requirement.
+        """
+        email = self._get_scope_value(result, "assume:mozilla-user:")
+        if not email:
```
Comment (djmitche) [inline treeherder/auth/backends.py]: You'll want to add `or email == '*'` here, otherwise all of us taskclusterites will get logged in as `*`.

File: treeherder/auth/backends.py (hunk 9 above)
Comment (camd) [inline treeherder/auth/backends.py]: good point, thanks.  done.

**Comment by djmitche** [comment]
/me unsubscribing from this thread -- ping me if you want me to take another look

**Comment by djmitche** [comment]
Actually, I think @camd has done one better -- scopes are being used directly to determine is_staff, not via an email address.

**Comment by edmorley** [comment]
That's what I mean - we shouldn't change that here, that's bug 1273092.

**Comment by edmorley** [comment]
ie preserve the current behaviour of using the DB's is_staff value (we don't use hardcoded emails for that now either).

File: ui/perf.html [hunk 12]
```diff
@@ -56,8 +55,9 @@
               ng-include="'partials/perf/helpMenu.html'">
           </ul>
         </span>
-        <!-- Persona login -->
-        <persona-buttons></persona-buttons>
+        <!-- login -->
+      <login on-user-change="$root.user = $event.user"></login>
```
Comment (wlach) [inline ui/perf.html]: This indentation looks wrong.

File: ui/js/components/auth.js [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+           ng-if="!user.loggedin"
+           ng-click="login()">Login/Register</a>
+    `,
+    bindings: {
+        onUserChange: "&"
+    },
+    controller: ['$scope', '$location', '$window', 'localStorageService',
+        'ThUserModel', '$http', 'thUrl', '$timeout',
+        function ($scope, $location, $window, localStorageService,
+                  ThUserModel, $http, thUrl, $timeout) {
```
Comment (wlach) [inline ui/js/components/auth.js]: I don't think you need to use $scope with new-style angular components, instead I think you can just attach things to `this`.  See `ui/js/components/perf/compare.js` for some examples.

File: ui/perf.html (hunk 12 above)
Comment (camd) [inline ui/perf.html]: heh, shoot.  how'd I manage that?  fixed.  :)

File: ui/js/components/auth.js (hunk 13 above)
Comment (camd) [inline ui/js/components/auth.js]: I took a look at your approach of ``ctrl = this`` and that seems reasonable enough.  I was using ``$scope`` so I could access the ``this`` type items in child functions and it seemed pretty familiar.  Since it's an isolate scope I didn't see it as much different.  But ``ctrl`` works well, too.  And it's more in line with the ``component`` pattern.   I had considered moving the anonymous functions out to be declared on ``this`` as well, but it just seemed like that would add more lines of code unnecessarily.  This feels like a good compromise.

File: treeherder/auth/backends.py [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+    def _get_email(self, result):
+        """
+        Get the user's email from the mozilla-user scope
+
+        """
+        email = self._get_scope_value(result, "assume:mozilla-user:")
+
+        if not email or email == "*":
+            # try finding the email in the clientId.  It can then be used to
```
Comment (edmorley) [inline treeherder/auth/backends.py]: This doesn't seem secure - we should fail authentication in `not email or email == "*"` case surely? I think it's reasonable to just say we don't support !LDAP people for now. (The description text on https://login.taskcluster.net/ says that non-employees can still use the Okta login option).

@djmitche, what are possible values for the `assume:mozilla-user:...` scope? Can a TC admin create arbitrary ones of these? (Is that what the `"*"` value is about?) Having to support (ie skip) missing email addresses is fine, but we absolutely do need to have a secure way to get a users email address, if it's present for a set of credentials.

Even with the previous version of the PR (that directly used the sheriff scope to set permissions), this would have been an issue, since we need to use the email address to migrate existing users. Admin/sheriff permission bits aren't the only sensitive items that need migrating, we also have Hawk API secrets in the Treeherder DB that are owned by individual users (that can't be represented by TC scopes). If someone can spoof the `assume:mozilla-user:...` scope, then they can view the Hawk secret for any API credentials owned by that user.

Comment (camd) [inline treeherder/auth/backends.py]: As I understand it, ``assume:mozilla-user:...`` will either be not present, have an email, or have * if the user is an admin.  So, yeah, that's the case this is trying to handle.  Dustin had talked about failing authentication in that case, thereby forcing the TC team to fix it on their end.  We can do that, for sure.

I'm not sure I follow your comment about "If someone can spoof the assume:mozilla-user:... scope".  If we can't trust the ``clientId`` and ``scopes`` that come back from the response, then we can't trust any of this anyway, right?   

If a user tries to use Persona, then they wouldn't have that ``mozilla-user`` scope.  But that's going away anyway, so not really an issue.

Though, you DO bring up a point that Dustin had brought up a while back and I'd forgotten about:  Trying to find a user with the truncated username (due to 30 char limit) could give the wrong user if the email address in their clientId was super long, making the trail end of it be the same for two different people. 

That being said, you're right.  We should just fail if we don't have a valid email from that scope.

File: treeherder/auth/backends.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+            raise ObjectDoesNotExist
+
+        # prefer a matching username.  If no username match, return the first
+        # user in the list.
+        try:
+            return user.get(username=username)
+        except ObjectDoesNotExist:
+            return user.first()
+
+    def authenticate(self, authorization=None, host=None, port=None):
```
Comment (edmorley) [inline treeherder/auth/backends.py]: Could you add a note saying that once Django 1.11 is released, and we're using it, we instead use the passed `request` object to determine the `host` and `port`:
https://docs.djangoproject.com/en/dev/topics/auth/customizing/#writing-an-authentication-backend

File: treeherder/auth/backends.py (hunk 14 above)
Comment (edmorley) [inline treeherder/auth/backends.py]: To clarify, I mean that I think we shouldn't fall back to using `result["clientId"]`, since Dustin said it was insecure?

File: treeherder/auth/backends.py [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+        user = None
+
+        if authorization:
+            tc_auth = Auth()
+            result = tc_auth.authenticateHawk({
```
Comment (edmorley) [inline treeherder/auth/backends.py]: Could you add a comment linking to https://docs.taskcluster.net/reference/platform/auth/api-docs#authenticateHawk please :-)

File: treeherder/auth/backends.py [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+                "authorization": authorization,
+                "host": host,
+                "port": port,
+                "resource": "/",
+                "method": "get",
+            })
+
+            if result["status"] == "auth-success":
```
Comment (edmorley) [inline treeherder/auth/backends.py]: How do you feel about converting this to an early return to reduce indentation?

File: treeherder/auth/backends.py [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+                # TODO: remove this size limit when we upgrade to django 1.10
+                # in Bug 1311967
+                username = result["clientId"][-30:]
```
Comment (edmorley) [inline treeherder/auth/backends.py]: Perhaps we should continue to use a hash or similar, to prevent someone's truncated email matching? (Or just blocking long email addresses - Gerard's long email was a gmail, I suspect he won't be able to use that now, since his Okta with be @mozilla.com)

Speaking of which, I guess we'll need to manually migrate people whose LDAP is not the same as their Persona email (only if they need specific sheriff/admin permissions or are credential owners). Happy to just do that post-deploy.

File: treeherder/auth/backends.py [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+                    logger.warning("Creating new user: {}".format(username))
+                    sha = sha1()
+                    sha.update(email)
+                    is_staff = scope_match(
+                        result["scopes"],
+                        [["assume:project:treeherder:sheriff"]]
+                    )
+                    user = User(email=email,
+                                username=username,
+                                password=sha.hexdigest()[-25:],
```
Comment (edmorley) [inline treeherder/auth/backends.py]: I think some of the users just have blank passwords at present (which I think means they are disabled). What happens if we just don't set a password here?

File: treeherder/auth/backends.py [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+        return user
+
+    def get_user(self, user_id):
+        try:
+            return User._default_manager.get(pk=user_id)
+        except User.DoesNotExist:
+            return None
+
+
+class TaskclusterAuthenticationFailed(Exception):
```
Comment (edmorley) [inline treeherder/auth/backends.py]: This seems unused?

File: treeherder/webapp/api/auth.py [hunk 21]
```diff
@@ -16,3 +27,28 @@ def hawk_lookup(id):
         'key': str(credentials.secret),
         'algorithm': 'sha256'
     }
+
+
+class TaskclusterAuthViewSet(viewsets.ViewSet):
+
+    @list_route(methods=['get'])
```
Comment (edmorley) [inline treeherder/webapp/api/auth.py]: `methods=['get']` is redundant, since it's the default (https://github.com/tomchristie/django-rest-framework/blob/3.3.3/rest_framework/decorators.py#L132), however if you'd prefer to include it for clarity, that's fine too.

File: treeherder/webapp/api/auth.py [hunk 22]
```diff
@@ -16,3 +27,28 @@ def hawk_lookup(id):
...
+    def login(self, request):
+        """
+        Verify credentials with Taskcluster
+
+        """
+        authorization = request.META.get("HTTP_OTH", None)
+        host = request.query_params.get("host", None)
```
Comment (edmorley) [inline treeherder/webapp/api/auth.py]: Could we extract these from `request` instead?

File: treeherder/auth/backends.py [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+                        [["assume:project:treeherder:sheriff"]]
+                    )
+                    user = User(email=email,
+                                username=username,
+                                password=sha.hexdigest()[-25:],
+                                is_staff=is_staff
+                                )
+
+                # update their email if their scopes when the user was created
+                # didn't allow finding the email, but now it's been fixed.
```
Comment (edmorley) [inline treeherder/auth/backends.py]: I think this comment is now out of date?

Do we want to update their username (away from the current hash value) instead?

File: treeherder/config/urls.py [hunk 24]
```diff
@@ -2,23 +2,19 @@
...
 from django.contrib import admin
-from django_browserid.admin import site as browserid_admin
 
 from treeherder.credentials.urls import urlpatterns as credentials_patterns
 from treeherder.embed import urls as embed_urls
 from treeherder.webapp.api import urls as api_urls
-from treeherder.webapp.views import LoginView
 
-browserid_admin.copy_registry(admin.site)
+admin.site.login_template = 'webapp/admin_login.html'
```
Comment (edmorley) [inline treeherder/config/urls.py]: Nice! I like that we can just get rid of the custom admin and replace with this template :-)
(I'm wondering if it wasn't possible to override the default template in the past, and that's why django-browserid did this)

File: treeherder/webapp/templates/webapp/admin_login.html [hunk 25]
```diff
@@ -0,0 +1,17 @@
...
+<html>
+  <body>
+    <div>
+        {% if user.is_authenticated %}
+        <!--shouldn't see this page if you're logged in.  It should just go to admin-->
+            Current user: {{ user.email }} - {{ user.username}} - {{user.is_superuser}}
+            <div>Something went wrong.
+                 Try logging out and back in through the main <a href="/" target="_blank">Treeherder UI</a>.</div>
+        {% else %}
+            <div>Please login through the main <a href="/" target="_blank">Treeherder UI</a>.</div>
```
Comment (edmorley) [inline treeherder/webapp/templates/webapp/admin_login.html]: This is a great interim solution - nice!

File: ui/js/components/auth.js [hunk 26]
```diff
@@ -0,0 +1,165 @@
...
+            var ctrl = this;
+            ctrl.user = {};
+            // "clears out" the user when it is detected to be logged out.
+            var loggedOutUser = {is_staff: false, username: "", email: "", loggedin: false};
+
+            /**
+             * Using a window listener here because I wasn't getting reliable
+             * results from the events from angular-local-storage.
+             */
+            $window.addEventListener("storage", function(e) {
```
Comment (edmorley) [inline ui/js/components/auth.js]: Ah cool - so multiple tabs will all pick up a login without needing a reload - great! (Does that even work at the moment with Persona?)

File: treeherder/config/urls.py (hunk 24 above)
Comment (camd) [inline treeherder/config/urls.py]: Yeah, that's quite possible.  We had a custom persona login template that I removed, but I actually think the template wasn't used at all.  It was just dead code, afaict.  :)

File: treeherder/auth/backends.py [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+    def _get_scope_value(self, result, scope_prefix):
+        for scope in result["scopes"]:
+            if scope.startswith(scope_prefix):
+                return scope[len(scope_prefix):]
+        return None
+
+    def _get_email(self, result):
+        """
+        Get the user's email from the mozilla-user scope
+
```
Comment (camd) [inline treeherder/auth/backends.py]: fixed

File: treeherder/auth/backends.py (hunk 19 above)
Comment (camd) [inline treeherder/auth/backends.py]: I testes this locally, and having a blank password doesn't seem to have any effect whatsoever.  So I'll just leave it out.  :)

File: treeherder/auth/backends.py (hunk 17 above)
Comment (camd) [inline treeherder/auth/backends.py]: I love it.  :)  done.

File: treeherder/auth/backends.py (hunk 20 above)
Comment (camd) [inline treeherder/auth/backends.py]: Using it now for when there's no email when they try to log in.

File: ui/js/components/auth.js (hunk 26 above)
Comment (camd) [inline ui/js/components/auth.js]: I doesn't work on Persona, no.  So this is an "upgrade"  :)

File: treeherder/auth/backends.py (hunk 23 above)
Comment (camd) [inline treeherder/auth/backends.py]: going with hash, per other comment.

File: ui/js/components/auth.js [hunk 28]
```diff
@@ -0,0 +1,165 @@
...
+            this.loginError = null;
+            const header = hawk.client.header(urlBase, 'GET', {
+                credentials: credentials,
+                ext: hawk.utils.base64urlEncode(JSON.stringify({"certificate": JSON.parse(certificate)}))}
+            );
+
+            // send a request from client side to TH server signed with TC
+            // creds from login.taskcluster.net
+            $http.get(`${urlBase}api/auth/login/`,
+                      {headers: {"oth": header.field}})
```
Comment (edmorley) [inline ui/js/components/auth.js]: Should we be POSTing the auth header to this endpoint instead? (Same for logout) Something doesn't seem right about making this a GET request, since it's not nullipotent. 

I'd also then prefer to either put the auth header in the body of the POST request, or else come up with another header name then "OTH". I'd say "Authorization" but I wonder if our own Hawk API middleware will grab it instead? (http://stackoverflow.com/questions/7802116/custom-http-authorization-header may also be of use)

File: ui/vendor/hawk/browser.js [hunk 29]
```diff
@@ -0,0 +1,793 @@
+'use strict';
+
+/*
+    HTTP Hawk Authentication Scheme
```
Comment (edmorley) [inline ui/vendor/hawk/browser.js]: There's no version number here - would be good to add one.
I'm guessing this might have also come from master?

Perhaps we should take from the latest tagged revision, and add a version number somewhere (either in the filename or else in this file comment)?
https://github.com/hueniverse/hawk/releases

File: ui/admin.html [hunk 30]
```diff
@@ -106,7 +106,6 @@
    <!--endbuild-->
 
   <script src="vendor/ngReact/ngReact.min.js"></script>
-  <script src="https://login.persona.org/include.js"></script>
```
Comment (edmorley) [inline ui/admin.html]: Yey for getting rid of the last third party JS include! Will make our CSP work simpler in the future :-)

File: treeherder/webapp/api/auth.py [hunk 31]
```diff
@@ -16,3 +28,39 @@ def hawk_lookup(id):
         'key': str(credentials.secret),
         'algorithm': 'sha256'
     }
+
+
+class TaskclusterAuthViewSet(viewsets.ViewSet):
```
Comment (edmorley) [inline treeherder/webapp/api/auth.py]: These API endpoints probably need some tests? (We'll need to mock out parts to prevent the requests to taskcluster login)

File: ui/js/components/auth.js [hunk 32]
```diff
@@ -0,0 +1,165 @@
...
+                            var storedUser = localStorageService.get("user");
+                            if (!storedUser || !storedUser.loggedin) {
+                                ctrl.setLoggedOut();
+                            }
+                        }
+                    }, 0);
+                }
+            });
+
+            // Ask the back-end if a user is logged in on page load
```
Comment (edmorley) [inline ui/js/components/auth.js]: This is great since it means if the Django session has expired since the last time the page was loaded (the session defaults to 2 weeks iirc), then we don't blindly use the user value from localstorage when they actually need to sign in again.

I was thinking it would probably be worth filing a new bug after this one, which moved this block to a separate function that could also be called by `ThModelErrors`, such that if we ever get a 401/403 from the API, we can double check whether they're still signed in, and update the state if they're not. (Currently I think they have to refresh the page manually).

File: ui/js/components/auth.js [hunk 33]
```diff
@@ -0,0 +1,165 @@
...
+            const port = $location.port();
+            const urlBase = `${$location.protocol()}://${host}:${port}/`;
+            const certificate = $location.search().certificate;
+            const credentials = {
+                id: $location.search().clientId,
+                key: $location.search().accessToken,
+                algorithm: 'sha256'
+            };
+            this.loginError = null;
+            const header = hawk.client.header(urlBase, 'GET', {
```
Comment (edmorley) [inline ui/js/components/auth.js]: Should we be Hawk signing `urlBase` or the full login URL? I think the latter seems more accurate? ie: The backend would then use a `resource` of "/api/auth/login/" not "/".

File: treeherder/auth/backends.py [hunk 34]
```diff
@@ -0,0 +1,123 @@
...
+                    'assume:project:ateam:*',
+                    'assume:project:treeherder:*',
+                    'assume:worker-id:*',
+                    'secrets:set:project/treeherder/*'],
+         'scheme': 'hawk',
+         'clientId': 'mozilla-ldap/biped@mozilla.com',
+         'expires': '2016-10-31T17:40:45.692Z'}
+    """
+
+    def _get_scope_value(self, result, scope_prefix):
```
Comment (camd) [inline treeherder/auth/backends.py]: yeah, good call.

File: treeherder/auth/backends.py [hunk 35]
```diff
@@ -0,0 +1,123 @@
...
+        for scope in result["scopes"]:
+            if scope.startswith(scope_prefix):
+                return scope[len(scope_prefix):]
+        return None
+
+    def _get_email(self, result):
+        """Get the user's email from the mozilla-user scope"""
+        email = self._get_scope_value(result, "assume:mozilla-user:")
+
+        if not email or email == "*":
```
Comment (camd) [inline treeherder/auth/backends.py]: sure, used the regex instead.  I'll see if I can sprinkle some more doc links around.  :)

File: treeherder/auth/backends.py [hunk 36]
```diff
@@ -0,0 +1,123 @@
...
+            raise TaskclusterAuthenticationFailed(
+                "Invalid email for user {} from scope 'assume:mozilla-user': {}".format(
+                    result["clientId"], email))
+        return email
+
+    def _get_user(self, email):
+        """
+        Try to find an exising user that matches either the username
+        or email.  Prefer the username, since that's the unique key.  But
+        fallback to the email.
```
Comment (camd) [inline treeherder/auth/backends.py]: fixed

File: treeherder/auth/backends.py [hunk 37]
```diff
@@ -0,0 +1,123 @@
...
+        """
+
+        # Since there is a unique index on username, but not on email,
+        # it is POSSIBLE there could be two users with the same email and
+        # different usernames.  Not very likely, but this is safer.
+        user = User.objects.filter(email=email)
```
Comment (camd) [inline treeherder/auth/backends.py]: added.

File: tests/auth/test_backends.py [hunk 38]
```diff
@@ -0,0 +1,58 @@
...
+
+from treeherder.auth.backends import TaskclusterAuthBackend
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize(('username', 'email', 'user', 'exp_exception'), [
+    ('dude', 'dude@lebowski.net', {"username": "dude", "email": "dude@lebowski.net"}, False),
+    ('thneed', 'dude@lebowski.net', {"username": "dude", "email": "dude@lebowski.net"}, False),
+    ('thneed', 'dood@lebowski.net', {"username": "dude", "email": "dude@lebowski.net"}, True)])
+def test_get_user(username, email, user, exp_exception):
```
Comment (camd) [inline tests/auth/test_backends.py]: fixed

File: treeherder/auth/backends.py [hunk 39]
```diff
@@ -0,0 +1,123 @@
...
+
+        if result["status"] != "auth-success":
+            return None
+
+        # TODO: remove this size limit when we upgrade to django 1.10
+        # in Bug 1311967
+        email = self._get_email(result)
+        sha = sha1()
+        sha.update(email)
+        username = sha.hexdigest()[-30:]
```
Comment (camd) [inline treeherder/auth/backends.py]: sure, that works for me.  :)

File: treeherder/auth/backends.py [hunk 40]
```diff
@@ -0,0 +1,123 @@
...
+            user = User(email=email,
+                        username=username,
+                        is_staff=is_staff
+                        )
+
+        # update the user object in the DB (perhaps Sheriff status
+        # or email changed.  This is so that when ``get_user`` is
+        # called, it will return the latest is_staff value we got from
+        # LDAP.
+        user.save()
```
Comment (camd) [inline treeherder/auth/backends.py]: left it since the <= 30 char clientId could be used now.

File: ui/js/components/auth.js [hunk 41]
```diff
@@ -0,0 +1,165 @@
...
+             */
+            ctrl.logout = function () {
+                $http.get(thUrl.getRootUrl("/auth/logout/"));
+                ctrl.setLoggedOut();
+            };
+
+            ctrl.setLoggedIn = function(newUser) {
+                newUser.loggedin = true;
+                localStorageService.set("user", newUser);
+                _.extend(ctrl.user, newUser);
```
Comment (camd) [inline ui/js/components/auth.js]: good point.  an earlier implementation needed that, but we don't anymore.

File: ui/js/components/auth.js (hunk 28 above)
Comment (camd) [inline ui/js/components/auth.js]: Yeah, the "Authorization" header was what I'd tried at first, and you're exactly right.  It got confused with the Hawk middleware.  But, sure, I'll make this a POST instead.  Makes sense.

File: treeherder/auth/backends.py [hunk 42]
```diff
@@ -85,9 +100,12 @@ def authenticate(self, authorization=None, host=None, port=None):
         # TODO: remove this size limit when we upgrade to django 1.10
         # in Bug 1311967
         email = self._get_email(result)
-        sha = sha1()
-        sha.update(email)
-        username = sha.hexdigest()[-30:]
+        if len(result["clientId"]) <= 30:
+            username = result["clientId"]
```
Comment (edmorley) [inline treeherder/auth/backends.py]: We should still always use the email for the username rather than `result["clientId"]`, the difference is that we only have to hash that email if it's >30 characters :-)

File: treeherder/webapp/api/auth.py [hunk 43]
```diff
@@ -16,3 +28,39 @@ def hawk_lookup(id):
...
+    def login(self, request):
+        """
+        Verify credentials with Taskcluster
+        """
+        authorization = request.META.get("HTTP_OTH", None)
+
+        # when we switch to Django 1.10, use ``get_host`` and ``get_port``
+        hostparts = request.META["HTTP_HOST"].split(":")
+        host = hostparts[0]
+        if "HTTP_X_FORWARDED_PORT" in request.META:
```
Comment (camd) [inline treeherder/webapp/api/auth.py]: ok, cool.  I did this.  It worked fine without changing anything in varnish both on stage and locally, so I guess we're ok there.  Had to add ``USE_X_FORWARDED_PORT = True`` to settings, though.

File: treeherder/webapp/api/auth.py [hunk 44]
```diff
@@ -16,3 +28,39 @@ def hawk_lookup(id):
...
+    @list_route()
+    def login(self, request):
+        """
+        Verify credentials with Taskcluster
+        """
+        authorization = request.META.get("HTTP_OTH", None)
+
+        # when we switch to Django 1.10, use ``get_host`` and ``get_port``
+        hostparts = request.META["HTTP_HOST"].split(":")
+        host = hostparts[0]
```
Comment (camd) [inline treeherder/webapp/api/auth.py]: good point.  done.

File: treeherder/auth/backends.py [hunk 45]
```diff
@@ -0,0 +1,123 @@
...
+        # see: https://docs.taskcluster.net/reference/platform/auth/api-docs#authenticateHawk
+        result = tc_auth.authenticateHawk({
+            "authorization": authorization,
+            "host": host,
+            "port": port,
+            "resource": "/",
+            "method": "get",
+        })
+
+        if result["status"] != "auth-success":
```
Comment (camd) [inline treeherder/auth/backends.py]: done

File: ui/js/components/auth.js (hunk 28 above)
Comment (camd) [inline ui/js/components/auth.js]: Talked on Vidyo and decided to stay with "GET" for now, but make the header be better named as "TC_AUTH"

File: ui/js/components/auth.js (hunk 33 above)
Comment (camd) [inline ui/js/components/auth.js]: ok, done.

File: ui/js/components/auth.js [hunk 46]
```diff
@@ -0,0 +1,165 @@
...
+        onUserChange: "&"
+    },
+    controller: ['$location', '$window', 'localStorageService',
+        'ThUserModel', '$http', 'thUrl', '$timeout',
+        function ($location, $window, localStorageService,
+                  ThUserModel, $http, thUrl, $timeout) {
+            var ctrl = this;
+            ctrl.user = {};
+            // "clears out" the user when it is detected to be logged out.
+            var loggedOutUser = {is_staff: false, username: "", email: "", loggedin: false};
```
Comment (camd) [inline ui/js/components/auth.js]: sure, we could use one of those instead.  But this is in lots of different files.  I'd rather fix that in a later PR, if that's cool.  I opened https://bugzilla.mozilla.org/show_bug.cgi?id=1318101 for this.

File: ui/js/components/auth.js [hunk 47]
```diff
@@ -0,0 +1,173 @@
...
+
+            // check if the user can login.  thServiceDomain must match
+            // host domain.  Remove this is we fix
+            // Bug 1317752 - Enable logging in with Taskcluster Auth cross-domain
+            var a = document.createElement('a');
+            a.href = thServiceDomain;
+            this.userCanLogin = a.hostname === $location.host();
```
Comment (edmorley) [inline ui/js/components/auth.js]: Could this just check to see if `thServiceDomain` is the empty string? (iirc it's empty when `local.conf.js` isn't being used). (This would save having to create an element)

File: ui/js/components/auth.js [hunk 48]
```diff
@@ -0,0 +1,173 @@
...
+                // open a new tab to show the taskcluster auth login page
+                $window.open(url, "_blank");
+            };
+
+            /**
+             * Contact Treeherder back-end to log the user out and invalidate
+             * the session token.  Then updates the UI
+             */
+            ctrl.logout = function () {
+                $http.get(thUrl.getRootUrl("/auth/logout/"));
```
Comment (edmorley) [inline ui/js/components/auth.js]: If the logout call fails (bad connection etc), then no error message will be shown, and the UI will erroneously show the user as logged out, until next page load.

File: ui/js/models/error.js [hunk 49]
```diff
@@ -7,7 +7,7 @@ message handling.
 treeherder.factory('ThModelErrors', [function() {
     // Generic error message when we encounter 401 status codes from the
     // server.
-    var AUTH_ERROR_MSG = 'Please login to Treeherder using Persona to ' +
+    var AUTH_ERROR_MSG = 'Please login to Treeherder to ' +
```
Comment (edmorley) [inline ui/js/models/error.js]: Perhaps squeeze this on one line?
